app/scripts/services/authorization.js

```diff
@@ -0,0 +1,121 @@
+'use strict';
+
+angular.module("openshiftConsole")
+  .factory("AuthorizationService", function($q, Logger, $interval, DataService){
+    function AuthorizationService() {
+      this.currentProject = null;
+      this.projects = {};
+      this.MAX_PROJECTS = 10;
+    }
+
+    var normalizeRules = function(rules) {
+      var normalizedRules = {};
+      var canAddToProject = null;
+      _.each(rules, function(rule) {
+        _.each(rule.resources, function(resource) {
+          normalizedRules[resource] = {
+            verbs: rule.verbs,
+            apiGroups: rule.apiGroups
+          };
+        });
+        if (_.contains(rule.verbs, "create") || _.contains(rule.verbs, "update")) {
+          canAddToProject = true;
+        }
+      });
+
+      return [normalizedRules, canAddToProject];
+    };
+
+    var getOldestProject = function(allProjectsRules) {
+      var oldestProject = {
+        name: null,
+        timestamp: _.now()
+      };
+      _.each(allProjectsRules, function(projectData, projectName){
+        if ((projectData.cacheTimestamp - oldestProject.timestamp) < 0) {
+          oldestProject.name = projectName;
+          oldestProject.timestamp = projectData.cacheTimestamp;
+        }
+      });
+      return oldestProject.name;
+    };
+
+    AuthorizationService.prototype.getProjectRules = function(projectName) {
+      var self = this;
+      var deferred = $q.defer();
+      this.currentProject = projectName;
+      if (!_.has(self.projects, projectName)) {
+        Logger.log("AuthorizationService, loading user rules for " + projectName + " project");
+
+        // Remove oldest project rules that we cache
+        if (_.size(self.projects) === self.MAX_PROJECTS) {
+          // Switch for `_.minBy` when we start to use Lodash 4.x.x
+          delete self.projects[getOldestProject(self.projects)];
+        }
+
+        var object = {kind: "SelfSubjectRulesReview",
+                      apiVersion: "v1"
+                    };
+        DataService.create('selfsubjectrulesreviews', null, object, {namespace: projectName}).then(
+          function(data) {
+            var normalizedData = normalizeRules(data.status.rules);
+            self.projects[projectName] = {
+              rules: normalizedData[0],
+              canAddToProject: normalizedData[1],
+              intervalTimer: $interval(function() {
+                                delete self.projects[projectName];
+                              }, 600000),
+              cacheTimestamp: _.now()
+            };
+
+            deferred.resolve(true);
+          }, function() {
+            deferred.reject(false);
+        });
+      } else {
+        // Using cached data.
+        // Just refresh the cacheTimestamp value to a current time. Not the interval.
+        Logger.log("AuthorizationService, using cached rules for " + projectName + " project");
```

Comment:
Instead of https://github.com/openshift/origin-web-console/pull/82/files#diff-493e422fe43d2ad97883bd51eeef2c38R66
Check the cacheTimestamp here and if its older than X amount of time kick off a background refresh of the rules and return what we have already cached.